cpp/src/arrow/compute/key_hash_internal.h

```diff
@@ -48,6 +48,12 @@ class ARROW_EXPORT Hashing32 {
   static void HashMultiColumn(const std::vector<KeyColumnArray>& cols, LightContext* ctx,
                               uint32_t* out_hash);
 
+  // Clarify the max temp stack usage for HashBatch so the caller could reserve enough
+  // size in advance.
```

Comment:
I'm not sure I understand this comment.

Reply:
Sorry for the confusion. This comment along with the following constant is because of the fact that `HashBatch` assumes an already-allocated stack rather than having its own, so the caller (the stack owner) may need to know it (possibly at compile time) in advance - otherwise the caller would have to be aware of the full implementation of `HashBatch`.

The case `HashBatchTempStackUsage` is to prevent the implantation of `HashBatch` from diverging with this constant.

The same idea applies to the rest of this PR.

Reply:
More comments added.